manual/src/tutorials/gadtexamples.etex

```diff
@@ -309,3 +309,119 @@ contradict the exhaustiveness check. Namely, builtin types (those
+impossible becaus the type "char t" is empty. This relies on the fact
+that the types occurring in the GADT index are known to be distinct:
+the constructor "Int" returns an "int t", and we know that "int" and
+"char" are distinct types, so "Int" is impossible at type "char t" --
+and similarly for "Bool". In other words, checking that a GADT case is
+impossible (at a specific type) requires reasoning on inequalities
+between types.
+
+For this reason, using abstract types as GADT indices does not work as
+intended. Consider for instance:
+
+\begin{caml_example*}{verbatim}
+(* type-level natural numbers, in unary notation *)
+module TypeNat = struct
+  type zero
+  type !'a succ
+end
+
+type ('a, _) sized_list =
+  | Nil : ('a, TypeNat.zero) sized_list
+  | Cons : 'a * ('a, 'n) sized_list -> ('a, 'n TypeNat.succ) sized_list
+
+type one = TypeNat.(zero succ)
+\end{caml_example*}
+
+This definition does not work: because "zero" and "succ" are abstract
+types, the compiler assumes that their definition is unknown, and that
+they might in fact be the same type. In particular, the types "zero"
+and "zero succ" are not known to be distinct, and the compiler cannot
+agree that the "Nil" case is impossible at type "zero
+succ":\footnote{We also needed an explicit injectivity annotation,
+  "type !'a succ", otherwise the definition of "sized_list" is also
+  rejected because "succ" is abstract.}
+\begin{caml_example}{verbatim}[error]
+let of_singleton : ('a, one) sized_list -> 'a = function
+  | Nil -> .
+  | Cons (v, Nil) -> v
+  | Cons (v, Cons _) -> .
+\end{caml_example}
+
+The recommended way to avoid this problem is to give a concrete
+definition to types such as "zero" and "succ", that are only meant to
+be used in GADT indices. (This is counter-intuitive, as we will never
+manipulate values of this type.) You could use:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = Zero
+  type 'a succ = Succ
+end
+\end{caml_example*}
+
+As an additional refinement, we can define those index types using
+using a "private" annotation, which clarifies that they are not meant
+to be used to build values:
+\begin{caml_example*}{verbatim}
+module TypeNat = struct
+  type zero = private Zero
+  type 'a succ = private Succ
+end
+\end{caml_example*}
+
+Remark: Usually "private" is meant to be used in interfaces, not in
+implementations; see the documentation of private types in
+\ref{s:private-types}. The constructors of a datatype that is
+\emph{defined} as "private" cannot be used to build values, so the
+type is effectively empty. This makes it more explicit that the types
+are only intended to be used as GADT indices.
+
+\subsection{s:abstract-types-current-module}{Legacy behavior: abstract types in the current module}
+
+Early versions of OCaml, up to OCaml 5.3, had a special handling of
+abstract types defined in the current module. (Not "succ" and "zero"
+above that are defined in a separate "TypeNat" submodule.) So the
```

Comment:
```suggestion
above that are defined in a separate "TypeNat" submodule). So the
```
If we follow the above proposal the period should also be moved outside the parens. :shrug: